core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java

```diff
@@ -0,0 +1,172 @@
+                }
+            }
+        } else {
+            sortValues = new Object[0];
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        if (sortValues.length > 0) {
+            builder.startArray(Fields.SORT);
+            for (Object sortValue : sortValues) {
+                builder.value(sortValue);
+            }
+            builder.endArray();
+        }
+        return builder;
+    }
+
+    public static SearchSortValues fromXContent(XContentParser parser) throws IOException {
+        assert parser.currentToken() == XContentParser.Token.FIELD_NAME;
+        assert parser.currentName().equals(Fields.SORT);
+        XContentParser.Token token = parser.nextToken();
+        assert token == XContentParser.Token.START_ARRAY;
+        return new SearchSortValues(parser.list().toArray());
+    }
+
+    public Object[] sortValues() {
+        return sortValues;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(sortValues.length);
+        if (sortValues.length > 0) {
+            for (Object sortValue : sortValues) {
+                if (sortValue == null) {
+                    out.writeByte((byte) 0);
+                } else {
+                    Class type = sortValue.getClass();
+                    if (type == String.class) {
+                        out.writeByte((byte) 1);
+                        out.writeString((String) sortValue);
+                    } else if (type == Integer.class) {
+                        out.writeByte((byte) 2);
+                        out.writeInt((Integer) sortValue);
+                    } else if (type == Long.class) {
+                        out.writeByte((byte) 3);
+                        out.writeLong((Long) sortValue);
+                    } else if (type == Float.class) {
+                        out.writeByte((byte) 4);
+                        out.writeFloat((Float) sortValue);
+                    } else if (type == Double.class) {
+                        out.writeByte((byte) 5);
+                        out.writeDouble((Double) sortValue);
+                    } else if (type == Byte.class) {
+                        out.writeByte((byte) 6);
+                        out.writeByte((Byte) sortValue);
+                    } else if (type == Short.class) {
+                        out.writeByte((byte) 7);
+                        out.writeShort((Short) sortValue);
+                    } else if (type == Boolean.class) {
+                        out.writeByte((byte) 8);
+                        out.writeBoolean((Boolean) sortValue);
+                    } else {
+                        throw new IOException("Can't handle sort field value of type [" + type + "]");
+                    }
+                }
+            }
+        }
+    }
+
+    public boolean xContentEquals(ToXContentToBytes other) {
```

Comment:
I think I'd prefer to leave this out and do `assertEquals(this.toString(), other.toString())` in the test.